_specifications/specification-3-15.md

```diff
@@ -128,10 +128,10 @@ interface ResponseError<D> {
 	message: string;
 
 	/**
-	 * A Primitive or Structured value that contains additional
+	 * A primitive or structured value that contains additional
 	 * information about the error. Can be omitted.
 	 */
-	data?: D;
+	data?: string | number | boolean | number | array | object | null;

```

Comment:
Sorry if I have misunderstood.
I think `array` is not included TypeScript primitive types.
So I think this is invalid type definition.

http://www.typescriptlang.org/play/?ssl=16&ssc=46&pln=16&pc=51#code/JYOwLgpgTgZghgYwgAgEoQM4AcD2IMQCiUUOUyA3gFACQA9AFQO3IPICCyIArgLYBG0ZKAAmwBHDCgA5sjAALFNFLkwATywoFk5DgQJuJCCIB0LBnVoIcIiAC4ufQVADcVWo2Y1WHZBjBQMshYpABuwGIgsnB+8mRgyLYYCIFYUni6MHKKyMpkZt4WtLyYGHDS9n4BMm4eTOa+IcC8wFKhKGRVUNwIYIbGyKFwADbcWvI61uBwoBjIcCJi6SAjDaAwZLySwBlw-DjcCQpKJPnIAMJwIMiCui1gkKbmljQiknAA-A7+gVHIAD6OARCQH7HDDCBXAFA5zQuAkOBqaE4fgAKwgvWhPGGwzcAF8gA

Reply:
Also, `number` is there twice.

Reply:
As the LSP is JSON, the valid types must be JSON data types. `any` is more than that.
Maybe we can define:
```ts
type LSPAny = LSPObject | LSPArray | string | number | boolean | null;
type LSPObject = { [key:string]:LSPAny };
type LSPArray = LSPAny[];
```
and use it where needed.